docs/rfcs/032-compute-aux-files.md

```diff
@@ -0,0 +1,112 @@
+# Postgres aux files storage
+
+In the current Neon architecture, compute nodes can only persist data through WAL replication to storage nodes. This method covers most of the data that Postgres persists, but not all. Exceptions include:
+
+* Replication slots
+* Postgres statistics files
+* `pg_stat_statements`` disk state
+* `pg_prewarm` disk state
+
+It may also be beneficial to store observability data, such as compute metrics and page access histograms, in the per-endpoint storage.
+
+## Aux file types
+
+The mentioned files have different requirements regarding durability, exhibit different I/O patterns, vary in expected size, and have different desired behaviors with respect to point-in-time recovery and read-only replicas.
+
+### Replication Slots
+
+Replication slots persist data about the replication receiver on the replication source. In the case of chain replication, these files could be modified by a replica (because it may have another attached replica). The ability to modify this state on a replica is the main reason for this data to be kept in a separate file instead of in the usual relation (replicas can't modify data).
```

Comment:
First of all it is not only slots.
We are persisting three kind of files related with logical replication:
1. slot state
2. snapshot
3. origins

Frankly speaking I do not think that this information is very different from the files we want to persists (stats info, shared buffer state, may be extension data...) so that it is necessary to create service just for serving slots.

> It's a small amount of data, could be stored in the control plane db or the db we use to store the generation information in storage now.

If we have 10k active tenants, each of which is generating snapshot each 15 seconds, then it is ~700 requests per second. Not so small workload. Typical snapshot is small (~150 nbytes) but if there is large number of active transactions, it can be easily as large as dozen of kilobytes.  It is megabytes per second. It is not a small amount of data.

Reply:
> Frankly speaking I do not think that this information is very different from the files we want to persists (stats info, shared buffer state, may be extension data...) so that it is necessary to create service just for serving slots.

Strong disagree. These are very different from other kinds of data you listed. Also, slot state, snapshots and origins are very different from _each other_.

Reply:
> If we have 10k active tenants, each of which is generating snapshot each 15 seconds, then it is ~700 requests per second. Not so small workload. Typical snapshot is small (~150 nbytes) but if there is large number of active transactions, it can be easily as large as dozen of kilobytes. It is megabytes per second. It is not a small amount of data.

- That sounds totally manageable by a single node. 
- That's a high estimate, I believe the average will be much less. Not all tenants use logical replication, and most won't have a large number of active transactions.
- The service can be easily sharded by tenant id if necessary.

Reply:
One more argument against  "slot synchronization service": slot manipulation should be consistent with state of the database. If we advance slot and this information is persisted, then Postgres is able to perform some cleanup, i.e. truncate CLOG, remove old versions (vacuum). If then we find out that this slot modification is lost, then we are in trouble and not able to continue replication any more. So we need conformation from the service that information is really persisted. And service should provide durability.

Right now we have three mechanism of enforcing durability: safekeepers, S3 and console plane Postgres database (RDS? or Aurora?).
If this service needs to save slot data in S3 and backend is blocked until it receives confirmation that this operation is completed then it may dramatically slowdown compute (although it is not in user's backend but walsender).

Relying in control plane database is IMHO not acceptable due to scalability issues.

Also there was concerns in this discussion that we should not rely on third-party solutions.
Isn't Aurora used by control plane an example of such third party solution?

Reply:
> One more argument against "slot synchronization service": slot manipulation should be consistent with state of the database. If we advance slot and this information is persisted, then Postgres is able to perform some cleanup, i.e. truncate CLOG, remove old versions (vacuum). If then we find out that this slot modification is lost, then we are in trouble and not able to continue replication any more. So we need conformation from the service that information is really persisted. And service should provide durability.

+1. Yes, we need to be careful with these properties in whatever mechanism we choose.

> Right now we have three mechanism of enforcing durability: safekeepers, S3 and console plane Postgres database (RDS? or Aurora?). If this service needs to save slot data in S3 and backend is blocked until it receives confirmation that this operation is completed then it may dramatically slowdown compute (although it is not in user's backend but walsender).

Sure, let's not do that.

> Relying on control plane database is IMHO not acceptable due to scalability issues.

Can you elaborate? But sure, it doesn't need to be the control plane database. It can be a different database dedicated to the service. Or shared with the 'generations' service.

> Also there was concerns in this discussion that we should not rely on third-party solutions. Isn't Aurora used by control plane an example of such third party solution?

That seems like a separate discussion. We already have a control plane database, so any concerns with relying on 3rd party services apply to that too, and any solutions to those concerns will apply too.

Reply:
> Can you elaborate? But sure, it doesn't need to be the control plane database. It can be a different database dedicated to the service. Or shared with the 'generations' service.

It should be  not just normal (standalone) Postgres, but Postgres with some high-availability warranties: RDS, Aurora... And I am not sure that it will not be a bottleneck. Certainly not now (when we have 1.5 clients using logical replication), but in future.

So let me clarify my point of view: I am not absolutely against dedicated service for persisting files. I just do not like an idea to use it only for replication slots and not for the files. I am not fully excited with solution with using logical replication messages for persisting files. But still think that it it the best compromise. It seems to be quite acceptable for persisting slots data and even more suitable for persisting statistic (which can benefit from versioning). There may be troubles with handling larger number of files or large files. But I do not consider them as slow stopper. 

So I think that developing separate service is overkill. At least at this moment. And in the future I still prefer to see something like DFS solution which will be able to handle arbitrary files (create by extensions).